### PR TITLE
add stream transformation support for linked events

### PR DESCRIPTION
**Pull request by MohamedFarouk-HMCTS**

### CODE REVIEW

File: stream-transformation-tool-service/src/test/java/uk/gov/justice/tools/eventsourcing/transformation/service/LinkedEventStreamTransformationServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import uk.gov.justice.services.eventsourcing.source.core.exception.EventStreamException;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class LinkedEventStreamTransformationServiceTest {
```
Comment (MartinYSpasov) [inline stream-transformation-tool-service/src/test/java/uk/gov/justice/tools/eventsourcing/transformation/service/LinkedEventStreamTransformationServiceTest.java]: This doesn't seem to test anything.

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java [hunk 2]
```diff
@@ -95,6 +99,9 @@ public void shouldCreateTasksForEachStream() {
 
         assertThat(startTransformation.outstandingTasks.size(), is(2));
         assertTrue(startTransformation.allTasksCreated);
+
+//        verify(linkedEventStreamTransformationService).truncateLinkedEvents();
+//        verify(linkedEventStreamTransformationService).populateLinkedEvents();
     }
```
Comment (allanmckenzie) [inline event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: Commented out code can be removed

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java [hunk 3]
```diff
@@ -119,6 +126,9 @@ public void shouldRemoveFinishedTasks() {
 
         startTransformation.taskDone(future2, null, null, null);
         assertThat(startTransformation.outstandingTasks.size(), is(0));
+
+//        verify(linkedEventStreamTransformationService).truncateLinkedEvents();
+//        verify(linkedEventStreamTransformationService).populateLinkedEvents();
     }
```
Comment (allanmckenzie) [inline event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: Commented out code can be removed

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java [hunk 4]
```diff
@@ -19,11 +20,11 @@
 
     private TestEventLogJdbcRepository eventLogJdbcRepository;
     private TestEventStreamJdbcRepository eventStreamJdbcRepository;
-
+    private EventFetcherRepository eventFetcherRepository = new EventFetcherRepository();
     private final LiquibaseUtil liquibaseUtil = new LiquibaseUtil();
-
+    DataSource dataSource = null;
```
Comment (allanmckenzie) [inline stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java]: Should be private (and final if possible)

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java [hunk 5]
```diff
@@ -19,11 +20,11 @@
 
     private TestEventLogJdbcRepository eventLogJdbcRepository;
     private TestEventStreamJdbcRepository eventStreamJdbcRepository;
-
+    private EventFetcherRepository eventFetcherRepository = new EventFetcherRepository();
```
Comment (allanmckenzie) [inline stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java]: final

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,9 @@ on [Keep a CHANGELOG](http://keepachangelog.com/). This project adheres to
 
 ## [Unreleased]
 
+## [5.2.0] - 2018-12-12
```
Comment (mapingo) [inline CHANGELOG.md]: looks like the wrong date

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java [hunk 7]
```diff
@@ -158,19 +168,28 @@ public void shouldRemoveFinishedTaskForAllPasses() {
...
 
 
     @Test(expected = IllegalArgumentException.class)
     public void shouldThrowExceptionWhenstreamsProcessedCountStepInfoIsNull() throws Exception {
         streamsProcessedCountStepInfo.set(startTransformation, null);
         startTransformation.go();
+//
+//        verify(linkedEventStreamTransformationService).truncateLinkedEvents();
+//        verify(linkedEventStreamTransformationService).populateLinkedEvents();
```
Comment (mapingo) [inline event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: this comment code should be reinstated or deleted

File: event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java [hunk 8]
```diff
@@ -174,6 +182,20 @@ private void shutdown() {
...
+        truncateAndPopulateLinkedEvents();
+    }
+
+    private void truncateAndPopulateLinkedEvents() {
+
+        linkedEventStreamTransformationService.truncateLinkedEvents();
+
+        logger.info("-------------- Truncating the Linked Events Log after complete transformation --------------");
+
+        linkedEventStreamTransformationService.truncateLinkedEvents();
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/main/java/uk/gov/justice/event/tool/StartTransformation.java]: removed

File: stream-transformation-test/sample-transformations/pom.xml [hunk 9]
```diff
@@ -23,5 +23,10 @@
             <artifactId>test-utils-core</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>uk.gov.justice.framework-api</groupId>
+            <artifactId>framework-api-core</artifactId>
+            <version>${framework-api.version}</version>
+        </dependency>
```
Comment (mapingo) [inline stream-transformation-test/sample-transformations/pom.xml]: If this is not a test dependency then it should be higher in the list

File: pom.xml [hunk 10]
```diff
@@ -23,8 +23,9 @@
     <properties>
         <cpp.repo.name>stream-transformation-tool</cpp.repo.name>
 
-        <framework.version>5.0.4</framework.version>
-        <event-store.version>1.0.4</event-store.version>
+        <framework.version>5.1.2-SNAPSHOT</framework.version>
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply pom.xml]: waiting for the release version of frame work api

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java]: brought it back in

File: stream-transformation-test/stream-transformation-it/pom.xml [hunk 11]
```diff
@@ -67,6 +67,20 @@
...
+        <dependency>
+            <groupId>uk.gov.justice.event-store</groupId>
+            <artifactId>linked-event-processor</artifactId>
+            <version>${event-store.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>uk.gov.justice.event-store</groupId>
+            <artifactId>event-repository-jdbc</artifactId>
+            <version>${event-store.version}</version>
+        </dependency>
```
Comment (mapingo) [inline stream-transformation-test/stream-transformation-it/pom.xml]: Non test dependencies above should be moved above the test dependency

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/EventLogBuilder.java [hunk 12]
```diff
@@ -15,7 +16,6 @@
     private EventLogBuilder() {
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/EventLogBuilder.java]: EventLogBuilder

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/TestLinkedEventJdbcRepository.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import javax.sql.DataSource;
+
+public class TestLinkedEventJdbcRepository extends LinkedEventJdbcRepository {
```
[the remark this answers is not in the document]
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/TestLinkedEventJdbcRepository.java]: yes

File: stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationMoveIT.java [hunk 14]
```diff
@@ -20,24 +21,25 @@
 public class StreamTransformationMoveIT {
 
     private static final UUID STREAM_ID = randomUUID();
+    private static final long STREAM_COUNT_REPORTING_INTERVAL = 10l;
```
Comment (mapingo) [inline stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationMoveIT.java]: uppercase L is easier to read - 10L

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java (hunk 5 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java]: added

File: stream-transformation-tool-service/src/main/java/uk/gov/justice/tools/eventsourcing/transformation/service/EventStreamTransformationService.java [hunk 15]
```diff
@@ -75,7 +75,7 @@ public UUID transformEventStream(final UUID originalStreamId, final int pass) {
 
             if (action.isTransform()) {
                 final Optional<UUID> newStreamId = eventTransformationStreamIdFilter.getEventTransformationStreamId(eventTransformations, jsonEnvelopeList);
-                if (newStreamId.isPresent()) {
+                if (newStreamId.isPresent() && !newStreamId.get().equals(originalStreamId)) {
```
Comment (mapingo) [inline stream-transformation-tool-service/src/main/java/uk/gov/justice/tools/eventsourcing/transformation/service/EventStreamTransformationService.java]: Can this check be created as a method or value and have a descriptive name?

File: CHANGELOG.md (hunk 6 above)
Comment (MohamedFarouk-HMCTS) [reply CHANGELOG.md]: fixed

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java (hunk 7 above)
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: done as mentioned by allan and martin

File: stream-transformation-test/sample-transformations/pom.xml (hunk 9 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/sample-transformations/pom.xml]: done

File: stream-transformation-test/stream-transformation-it/pom.xml (hunk 11 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/pom.xml]: done

File: stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationMoveIT.java (hunk 14 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/test/java/uk/gov/justice/framework/tools/transformation/StreamTransformationMoveIT.java]: Done

File: stream-transformation-tool-service/src/main/java/uk/gov/justice/tools/eventsourcing/transformation/service/EventStreamTransformationService.java (hunk 15 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-tool-service/src/main/java/uk/gov/justice/tools/eventsourcing/transformation/service/EventStreamTransformationService.java]: DONE

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java (hunk 3 above)
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: removed

File: stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java (hunk 4 above)
Comment (MohamedFarouk-HMCTS) [reply stream-transformation-test/stream-transformation-it/src/main/java/uk/gov/justice/framework/tools/transformation/DatabaseUtils.java]: private not final used outside constructor

File: event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java (hunk 2 above)
Comment (MohamedFarouk-HMCTS) [reply event-tool/src/test/java/uk/gov/justice/event/tool/StartTransformationTest.java]: removed

File: pom.xml [hunk 16]
```diff
@@ -22,9 +22,9 @@
 
     <properties>
         <cpp.repo.name>stream-transformation-tool</cpp.repo.name>
-
-        <framework.version>5.0.4</framework.version>
-        <event-store.version>1.0.4</event-store.version>
+        <framework-api.version>4.0.0-M2</framework-api.version>
```
Comment (mapingo) [inline pom.xml]: framework-api version is 3.1.1

File: pom.xml [hunk 17]
```diff
@@ -22,9 +22,9 @@
...
+        <framework.version>6.0.0-M4</framework.version>
```
Comment (mapingo) [inline pom.xml]: framework version is 5.1.2

File: CHANGELOG.md [hunk 18]
```diff
@@ -5,6 +5,9 @@ on [Keep a CHANGELOG](http://keepachangelog.com/). This project adheres to
 
 ## [Unreleased]
 
+## [5.2.0] - 2019-03-18
```
Comment (mapingo) [inline CHANGELOG.md]: Wrong date now

Comment (MohamedFarouk-HMCTS) [reply CHANGELOG.md]: fixed

File: pom.xml (hunk 17 above)
Comment (MohamedFarouk-HMCTS) [reply pom.xml]: fixed

File: pom.xml (hunk 16 above)
Comment (MohamedFarouk-HMCTS) [reply pom.xml]: fixed

**Pull request merged**